ext/opcache/zend_file_cache.c

```diff
@@ -1641,6 +1648,98 @@ static void zend_file_cache_unserialize(zend_persistent_script  *script,
 	UNSERIALIZE_PTR(script->arena_mem);
 }
 
+static int zend_opcache_check_opcode_file_tag(int fd, char* filename) {
+	zend_opcode_file_taginfo taginfo;
+	uint32_t php_version_id = (uint32_t)PHP_VERSION_ID;
+
+	int rn = read(fd, &taginfo, sizeof(zend_opcode_file_taginfo));
+	if(rn != sizeof(taginfo)) {
+		return -1;
+	}
+
+	if(memcmp(taginfo.tagname, T_PHP_OPCODE, T_PHP_OPCODE_LEN) != 0) {
+		return -1;
+	}
+
+	if(memcmp(&taginfo.php_version_id, &php_version_id, sizeof(php_version_id)) != 0) {
+		int version_errno = E_WARNING;
+		if(ZCG(accel_directives).prohibit_different_version_opcode) {
+			version_errno = E_ERROR;
+		} 
+		zend_error(version_errno, "Opcode file '%s' (version id: %d) is compiled by different versions of PHP (version id: %d), it's not safe and not suggest", filename, taginfo.php_version_id, php_version_id);
+	}
+	return 0;
+}
+
+void zend_opcache_copy_file_error(char* msg, char* file1, char *file2, int free) {
+	zend_error(E_WARNING, msg, file1);
+	zend_file_cache_unlink(file2);
+	zend_file_cache_unlink(file1);
+	if(free) { 
+		efree(file1);
+	} else {
+		efree(file2);
+	}
+}
+
+int zend_opcache_copy_opcode_cache_file(char *src_filename, size_t src_filename_len, char* opcode_file) {
+	zend_string *full_path;
+	char full_path_buff[MAXPATHLEN];
+	char *cache_file;
+	int fd_new;
+	int fd_cache;
+	char buf[8192];
+	int buf_len;
+	int chunksize = sizeof(buf);
+	int write_size = 0;
+
+	zend_opcode_file_taginfo taginfo = {.tagname=T_PHP_OPCODE, .php_version_id= (uint32_t)PHP_VERSION_ID};
+
+	VCWD_REALPATH(src_filename,full_path_buff);
+	full_path = zend_string_init(full_path_buff, strlen(full_path_buff), 0);
+	cache_file = zend_file_cache_get_bin_file_path(full_path);
+	efree(full_path);
+	fd_new = zend_file_cache_open(opcode_file, O_CREAT | O_EXCL | O_RDWR | O_BINARY, S_IRUSR | S_IWUSR);
+	if(fd_new < 0) {
+		zend_opcache_copy_file_error("create file '%s' error\n", opcode_file, cache_file, 0);
+		return FAILURE;
+	}
+
+	if(write(fd_new, &taginfo, sizeof(taginfo)) != sizeof(taginfo)) {
+		zend_opcache_copy_file_error("write taginfo to file '%s' error\n", opcode_file, cache_file, 0);
+		return FAILURE;
+	}
+
+	fd_cache = zend_file_cache_open(cache_file, O_RDONLY | O_BINARY);
+	if(fd_cache < 0) {
+		zend_opcache_copy_file_error("can not open opcode cache file '%s' error\n",cache_file, opcode_file, 1);
+		return FAILURE;
+	}
+
+	while(1) {
+		buf_len = read(fd_cache, buf, chunksize);
+		if(buf_len <= 0) {
+			break;
+		}
+		write_size = write(fd_new, buf, buf_len);
+		if(write_size != buf_len) {
+			zend_opcache_copy_file_error("write file '%s' error\n", opcode_file, cache_file, 0);
+			return FAILURE;
+		}
+		if(buf_len < chunksize) {
```

Comment:
> On success, the number of bytes read is returned (zero indicates end
> of file), and the file position is advanced by this number.  **It is
> not an error if this number is smaller than the number of bytes
> requested; this may happen for example because fewer bytes are
> actually available right now (maybe because we were close to end-of-
> file, or because we are reading from a pipe, or from a terminal), or
> because read() was interrupted by a signal.**  See also NOTES.

I'd suggest just deleting this if statement entirely

Reply:
* `if(buf_len < chunksize) ` 
It is guaranteed to exit at the end of the file
*  `read`
 There may be a buffer problem that causes an exception to read null character data